File: book/src/schema/enums/advanced.kt
fun main() {}

sealed class Sample {
    class Variant1(
        val f1: UInt,
        val f2: UShort,
    ) : Sample()

    class Variant2(
        val field1: Long,
        val field2: Boolean,
    ) : Sample()

    // N variants...
}

File: book/src/schema/enums/advanced.ts
enum SampleKind {
    Variant1,
    Variant2,
    // N variants...
}

interface SampleVariant1 {
    kind: SampleKind.Variant1;
    f1: number;
    f2: number;
}

interface SampleVariant2 {
    kind: SampleKind.Variant2;
    field1: number;
    field2: boolean;
}
